### PR TITLE
Media User & Detected Metadata

### PR DESCRIPTION
**Pull request by CarsonF**
Currently this only serves to enhance the existing file functionality.

With this scope I think the functionality is complete, save for:
- [ ] Allow user metadata to be updated after upload
- [ ] Migration to detect metadata for existing files

### CODE REVIEW

File: src/core/database/highlight-cypher.util.ts [hunk 1]
```diff
@@ -9,7 +9,15 @@ export function highlight(query: string) {
...
   });
-  pretty = pretty.split('\n').slice(0, -1).join('\n');
+  pretty = pretty
+    .split('\n')
+    .slice(0, -1)
+    // strip off trailing spaces meant for a "code block background."
+    // we don't set a background color, and this makes copy/pasting into
+    // neo4j editor painful.
+    // eslint-disable-next-line no-control-regex
+    .map((line) => line.replace(/( )+\u001B\[39m$/, ''))
```
Comment (brentkulwicki) [inline src/core/database/highlight-cypher.util.ts:19]: What is this pattern matching for here that a .trim() wouldn't work for?

**Review by brentkulwicki — APPROVED** [review]
One question but looks good overall.

File: src/core/database/highlight-cypher.util.ts (hunk 1 above)
Comment (CarsonF) [reply src/core/database/highlight-cypher.util.ts:19]: Trim doesn't work because the spaces are followed by that terminal escape sequence. I tried lol

File: src/components/file/media/media-detector.service.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        '-show_streams',
+        '-i',
+        'pipe:0',
+      ],
+      {
+        reject: false,
+        input: stream,
+      },
+    );
+    return JSON.parse(stdout);
```
Comment (adam-soltech) [inline src/components/file/media/media-detector.service.ts:81]: Since reject is set to false should there be some error handling here in case stdout ends up not being valid json?

File: src/components/file/media/media-detector.service.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      return {
+        __typename: 'Image',
+        dimensions: {
+          width: size.width ?? 0,
+          height: size.height ?? 0,
+        },
+      };
+    }
+
+    const isAudio = file.mimeType.startsWith('audio/');
```
Comment (adam-soltech) [inline src/components/file/media/media-detector.service.ts:36]: Will this sort of audio, video, image strategy be repeated anywhere else? IF so it might be worth assigning 'audio/', 'video/' etc to an enum rather than hardcoding them

File: src/components/file/media/media.repository.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+    const result = await query.first();
+    if (!result) {
+      throw new ServerException('Failed to save media info');
+    }
+    return result.dto;
+  }
+}
+
+const toDbShape = (input: Partial<AnyMedia>) => ({
```
Comment (adam-soltech) [inline src/components/file/media/media.repository.ts:131]: I think I groked this but it was a bit hard to follow....

Probably a bit of a nit but something like this might be easier to read:
```

const getDimensions = (typename: string | undefined, dimensions: { width?: number | null, height?: number | null }) => {
  return (typename === 'Image' || typename === 'Video') 
    ? dimensions 
    : { width: null, height: null };
};

const getDuration = (typename: string | undefined, duration: number | undefined) => {
  return { duration: typename === 'Audio' ? duration : null };
};

const toDbShape = (input: Partial<AnyMedia>) => {
  return {
    type: input.__typename,
    altText: input.altText,
    caption: input.caption,
    mimeType: input.mimeType,
    ...getDimensions(input.__typename, input.dimensions),
    ...getDuration(input.__typename, input.duration),
  };
};
```

**Review by adam-soltech — COMMENTED** [review]
Looks great overall. I had a few questions and suggestions but nothing blocking

File: src/components/file/media/media-detector.service.ts (hunk 2 above)
Comment (CarsonF) [reply src/components/file/media/media-detector.service.ts:81]: Yeah that's a good idea.

File: src/components/file/media/media-detector.service.ts (hunk 3 above)
Comment (CarsonF) [reply src/components/file/media/media-detector.service.ts:36]: I doubt it. This is the only spot we have logic based on mime types. But definitely something to keep in mind for future.
Depending on the future need, the media type could be check instead, which is a string literal union.
```ts
media.__typename === 'Audio'
```

File: src/components/file/media/media.repository.ts (hunk 4 above)
Comment (CarsonF) [reply src/components/file/media/media.repository.ts:131]: Part of why this is syntax heavy is the nested check for null `__typename`. I wanted this to be used for multiple purposes:
1. Saving detected metadata, and if for whatever reason type changes we want to clear out previous/stale values
2. Saving user metadata. Which should just need the `altText`, `caption` as input. So we'd want to ignore typename logic there and not clear out the existing values.

That additional check could certainly be added to your helper functions above, or some combination. But IMO they don't add enough value to warrant the additional typing needed for the function inputs.

**Comment by CarsonF** [comment]
Thanks for the reviews guys. The biggest concern with new features is the schema, so that's where I'd like the most consideration (just FYI). Internal bits can always be tweaked.

I'm not sure I'm thrilled about the naming here.
https://github.com/SeedCompany/cord-api-v3/blob/5eebbe18f6cefd72bb6594c8d1fde8f60f012079/src/components/file/dto/upload.dto.ts#L40-L40
Both the property name and the type name.

Same for property name here
```gql
interface FileNode {
  media: Media
```
Sometimes I'm too terse for my own good. I think I'm ok with this case. I don't think changing to `mediaInfo` for example would add more clarity.

Open to suggestions on all the additional naming added
